cursos/kubernetes/02_kubernetes/08_taints_and_tolerations.md

```diff
@@ -0,0 +1,148 @@
+  extraPortMappings:
+  - containerPort: 8080
+    hostPort: 8080
+    protocol: TCP
+  - containerPort: 4443
+    hostPort: 4443
+    protocol: TCP
+- role: worker
+  image: kindest/node:v1.26.0
+  kubeadmConfigPatches:
+  - |
+    kind: InitConfiguration
+    nodeRegistration:
+      kubeletExtraArgs:
+        node-labels: "ingress-ready=true"
+        node-labels: "worker=worker"
+        system-reserved: memory=1000m,cpu="1"
+- role: worker
+  image: kindest/node:v1.26.0
+  kubeadmConfigPatches:
+  - |
+    kind: InitConfiguration
+    nodeRegistration:
+      kubeletExtraArgs:
+        node-labels: "ingress-ready=true"
+        node-labels: "worker=worker"
+        system-reserved: memory=1000m,cpu="1"
+
+```
+
+Ejercicio: 
+
+a) Aplicar taints a un nodo y ver el comportamiento de los pods:
+
+b) Verificar que la taint se ha aplicado correctamente usando el comando kubectl describe node <node-name>.
+
+c) Crear un pod con una toleration que coincida con la taint usando el campo tolerations en el archivo YAML del pod.
+
+d) Verificar que el pod se ha programado en el nodo con la taint usando el comando kubectl get pods -o wide.
+
+Repetir los pasos anteriores con diferentes combinaciones de taints y tolerations para ver cómo afectan al comportamiento de los pods y los nodos.
+
+
+
+# PodAffinity y NodeAffinity
+
+Otra forma de controlar en que nodo se van a desplegar los pods es mediante el uso de podAffinity y nodeAffinity.
+
+[Pod affinity y node affinity](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/) son dos conceptos de Kubernetes que permiten asignar pods a nodos según ciertas condiciones. Pod affinity se refiere a la preferencia o requerimiento de que un pod se ejecute en el mismo nodo que otro pod existente. Node affinity se refiere a la preferencia o requerimiento de que un pod se ejecute en un nodo que tenga ciertas etiquetas o atributos.
+
+Ejemplo de podAffinity:
+
+```
+affinity:
+  podAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+    - labelSelector:
+        matchExpressions:
+        - key: name
+          operator: In
+          values:
+          - web-app
+      topologyKey: kubernetes.io/hostname
+    preferredDuringSchedulingIgnoredDuringExecution:
+    - weight: 100
+      podAffinityTerm:
+        labelSelector:
+          matchExpressions:
+          - key: type
+            operator: In
+            values:
+            - primary
+        topologyKey: kubernetes.io/hostname
+
+    nodeAffinity:  
+      requiredDuringSchedulingIgnoredDuringExecution: 
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: type
+            operator: NotIn 
+            values:
+            - primary
+  containers:
+  - name: node-affinity
+    image: nginx
+
+```
+
+Estas 2 practicas para controlar pods se pueden combinar para formar reglas mas precisas que controlen el funcionamiento de los pods en nodos de kubernetes
+
+
+
+
```

Comment:
```suggestion
```

Reply:
I would add an extra section on [topologySpreadConstraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) to spread pods equally over all nodes.

https://www.danielstechblog.io/distribute-your-application-across-different-availability-zones-in-aks-using-pod-topology-spread-constraints/